src/Vimeo/ABLincoln/Namespaces/NamespaceTrait.php

```diff
@@ -0,0 +1,310 @@
+     * @param LoggerInterface $logger optional PSR-3 logging instance to use
+     */
+    public function initialize($inputs, LoggerInterface $logger = null)
+    {
+        $this->inputs = $inputs;         // input data
+        $this->name = get_class($this);  // use class name as default name
+        $this->num_segments = null;      // num_segments set in setup()
+        $this->in_experiment = false;    // not in experiment until unit assigned
+        $this->logger = $logger;         // track PSR logger if assigned
+
+        // array mapping segments to experiment names
+        $this->segment_allocations = array();
+
+        // array mapping experiment names to experiment objects
+        $this->current_experiments = array();
+
+        $this->experiment = null;          // memoized experiment object
+        $this->default_experiment = null;  // memoized default experiment object
+        $this->default_experiment_class = 'Vimeo\ABLincoln\Experiments\DefaultExperiment';
+
+        // setup name, primary key, number of segments, etc
+        $this->setup();
+        $this->available_segments = range(0, $this->num_segments - 1);
+        
+        $this->setupExperiments();  // load namespace with experiments
+    }
+
+    /**
+     * Set namespace attributes for run. Developers extending this class should
+     * set the following variables:
+     *     this->name = 'sample namespace';
+     *     this->primary_unit = 'userid';
+     *     this->num_segments = 10000;
+     */
+    abstract public function setup();
+
+    /**
+     * Setup experiments segments will be assigned to:
+     *     $this->addExperiment('first experiment', Exp1, 100);
+     */
+    abstract public function setupExperiments();
+
+    /**
+     * Get the primary unit that will be mapped to segments
+     *
+     * @return array array containing value(s) used for unit assignment
+     */
+    public function primaryUnit()
+    {
+        return $this->primary_unit;
+    }
+
+    /**
+     * Set the primary unit that will be used to map to segments
+     *
+     * @param mixed $value value or array used for unit assignment to segments
+     */
+    public function setPrimaryUnit($value)
+    {
+        $this->primary_unit = is_array($value) ? $value : array($value);
+    }
+
+    /**
+     * In-experiment accessor
+     *
+     * @return boolean true if primary unit mapped to an experiment, false otherwise
+     */
+    public function inExperiment()
+    {
+        $this->_requiresExperiment();
+        return $this->in_experiment;
+    }
+
+    /**
+     * Map a new experiment to a given number of segments in the namespace
+     *
+     * @param string $name name to give the new experiment
+     * @param string $exp_class string version of experiment class to instantiate
+     * @param int $num_segments number of segments to allocate to experiment
+     */
+    public function addExperiment($name, $exp_class, $num_segments)
```

Comment:
Addressed in c5674f30 (additional random operator sanity checks added in af3527)